src/tool/subcommands/state_migration_cmd.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::cli::subcommands::car_cmd::merge_car_streams;
+use crate::networks::{ActorBundleInfo, ACTOR_BUNDLES};
+use crate::utils::db::car_stream::CarStream;
+use crate::utils::net::global_http_client;
+use anyhow::{Context as _, Result};
+use async_compression::futures::write::ZstdEncoder;
+use cid::Cid;
+use clap::Subcommand;
+use futures::io::{BufReader, BufWriter};
+use futures::{AsyncRead, AsyncWriteExt, StreamExt, TryStreamExt};
+use fvm_ipld_car::{CarHeader, CarReader};
+use itertools::Itertools;
+use once_cell::sync::Lazy;
+use reqwest::Url;
+use std::env;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+const DEFAULT_BUNDLE_FILE_NAME: &str = "actor_bundles.car.zst";
+
+static ACTOR_BUNDLE_CACHE_DIR: Lazy<PathBuf> =
+    Lazy::new(|| env::temp_dir().join(".forest_actor_bundles/"));
+
+#[derive(Debug, Subcommand)]
+pub enum StateMigrationCommands {
+    /// Generate a merged actor bundle in `.car.zst` format under the current working directory.
+    ActorBundle,
+}
+
+impl StateMigrationCommands {
+    pub async fn run(self) -> Result<()> {
+        use StateMigrationCommands::*;
+
+        match self {
+            ActorBundle => generate_actor_bundle().await,
+        }
+    }
+}
+
+async fn generate_actor_bundle() -> Result<()> {
+    let mut tasks = Vec::with_capacity(ACTOR_BUNDLES.len());
+    for ActorBundleInfo { manifest, url } in ACTOR_BUNDLES.iter() {
+        tasks.push(tokio::spawn(async move {
+            download_bundle_if_needed(manifest, url)
+                .await
+                .with_context(|| format!("Failed to get {manifest}.car from {url}"))
+        }));
+    }
+
+    let mut car_streams = vec![];
+    for task in tasks {
+        let path = task.await??;
+        let car_stream = CarStream::new(tokio::io::BufReader::new(
+            tokio::fs::File::open(path).await?,
+        ))
+        .await?;
+        car_streams.push(car_stream);
+    }
+
+    let all_roots = car_streams
+        .iter()
```

Comment:
Not sure if there's anything to be gained in parallelising this. Could be interesting depending on the times it takes to run this.

Reply:
roots are part of CAR header which has been parsed in `CarStream::new` so this call super fast